### ISSUE TITLE
The training got stuck and there is no progress.

### ISSUE THREAD

**Issue by LHB-kk** [zxiangx/LC-R1#2]
Hello, I am trying to reproduce the experiment, but I found that the training got stuck and there is no progress.

```sh
[2025-07-03 13:32:40,450] [INFO] [config.py:925:print]   zero_enabled ................. False
[2025-07-03 13:32:40,450] [INFO] [config.py:925:print]   zero_force_ds_cpu_optimizer .. True
[2025-07-03 13:32:40,450] [INFO] [config.py:925:print]   zero_optimization_stage ...... 0
[2025-07-03 13:32:40,450] [INFO] [config.py:911:print_user_config]   json = {
    "train_batch_size": 48, 
    "train_micro_batch_size_per_gpu": 2, 
    "gradient_accumulation_steps": 8, 
    "zero_optimization": {
        "stage": 0, 
        "offload_optimizer": {
            "device": "cpu", 
            "nvme_path": null
        }, 
        "offload_param": {
            "device": "cpu", 
            "nvme_path": null
        }, 
        "stage3_gather_16bit_weights_on_model_save": false
    }, 
    "gradient_clipping": 0.1, 
    "steps_per_print": inf, 
    "bf16": {
        "enabled": true
    }, 
    "fp16": {
        "enabled": false
    }
}
Gradient accumulation steps mismatch: GradientAccumulationPlugin has 1, DeepSpeed config has 8. Using DeepSpeed's value.
Using /root/.cache/torch_extensions/py311_cu124 as PyTorch extensions root...
Detected CUDA files, patching ldflags
Emitting ninja build file /root/.cache/torch_extensions/py311_cu124/cpu_adam/build.ninja...
/usr/local/lib/python3.11/site-packages/torch/utils/cpp_extension.py:2059: UserWarning: TORCH_CUDA_ARCH_LIST is not set, all archs for visible cards are included for compilation. 
If this is not desired, please set os.environ['TORCH_CUDA_ARCH_LIST'].
  warnings.warn(
Building extension module cpu_adam...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
ninja: no work to do.
Loading extension module cpu_adam...
Time to load cpu_adam op: 2.2426810264587402 seconds
Using /root/.cache/torch_extensions/py311_cu124 as PyTorch extensions root...
Detected CUDA files, patching ldflags
Emitting ninja build file /root/.cache/torch_extensions/py311_cu124/cpu_adam/build.ninja...
/usr/local/lib/python3.11/site-packages/torch/utils/cpp_extension.py:2059: UserWarning: TORCH_CUDA_ARCH_LIST is not set, all archs for visible cards are included for compilation. 
If this is not desired, please set os.environ['TORCH_CUDA_ARCH_LIST'].
  warnings.warn(
Building extension module cpu_adam...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
ninja: no work to do.
Loading extension module cpu_adam...
Time to load cpu_adam op: 2.2484261989593506 seconds
Using /root/.cache/torch_extensions/py311_cu124 as PyTorch extensions root...
Detected CUDA files, patching ldflags
Emitting ninja build file /root/.cache/torch_extensions/py311_cu124/cpu_adam/build.ninja...
/usr/local/lib/python3.11/site-packages/torch/utils/cpp_extension.py:2059: UserWarning: TORCH_CUDA_ARCH_LIST is not set, all archs for visible cards are included for compilation. 
If this is not desired, please set os.environ['TORCH_CUDA_ARCH_LIST'].
  warnings.warn(
Building extension module cpu_adam...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
ninja: no work to do.
Loading extension module cpu_adam...
Time to load cpu_adam op: 2.2451608180999756 seconds
  0%|                                                                                                                                                                | 0/417 [00:00<?, ?it/s
````

the vLLM log:
```sh
INFO:     127.0.0.1:59174 - "POST /update_named_param/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59174 - "POST /update_named_param/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59174 - "POST /update_named_param/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59174 - "POST /update_named_param/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59174 - "POST /update_named_param/ HTTP/1.1" 200 OK
INFO 07-03 13:33:05 [block_pool.py:264] Successfully reset prefix cache
INFO:     127.0.0.1:59174 - "POST /reset_prefix_cache/ HTTP/1.1" 200 OK

Processed prompts:   0%|          | 0/48 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]
Processed prompts:  17%|█▋        | 8/48 [00:46<03:51,  5.79s/it, est. speed input: 7.77 toks/s, output: 398.63 toks/s]
Processed prompts:  33%|███▎      | 16/48 [01:36<03:13,  6.06s/it, est. speed input: 11.05 toks/s, output: 562.18 toks/s]
Processed prompts:  50%|█████     | 24/48 [01:48<01:35,  3.98s/it, est. speed input: 13.43 toks/s, output: 888.01 toks/s]
Processed prompts:  67%|██████▋   | 32/48 [01:53<00:42,  2.64s/it, est. speed input: 18.04 toks/s, output: 1193.96 toks/s]
Processed prompts: 100%|██████████| 48/48 [01:54<00:00,  1.29s/it, est. speed input: 27.75 toks/s, output: 1936.83 toks/s]
Processed prompts: 100%|██████████| 48/48 [01:54<00:00,  2.38s/it, est. speed input: 27.75 toks/s, output: 1936.83 toks/s]
INFO:     127.0.0.1:59174 - "POST /generate/ HTTP/1.1" 200 OK
```

And my script:
```sh
#!/bin/bash
MODEL_PATH=Qwen3-4B # replace to your own model path
SAVE_PATH=qwen3-4B-dpo  # replace to your own save path
CONFIG_PATH=accelerate_configs/deepspeed_zero2.yaml

mkdir -p "$SAVE_PATH"
mkdir -p log

# run vllm-serve
export CUDA_VISIBLE_DEVICES=3
fuser -k 8000/tcp
trl vllm-serve --model "$MODEL_PATH" --port 8000 > log/vllm.log 2>&1 &
VLLM_PID=$!
echo "vLLM server PID: $VLLM_PID"

accelerate launch --config_file $CONFIG_PATH \
    train.py \
    --save_path "$SAVE_PATH" \
    --model_path "$MODEL_PATH" \
    --lr 3e-6 \
    --num_generations 8\
    --num_train_epochs 1\
    --run_name lcr1-qwen3-4b-grpo-0702\
    --per_device_train_batch_size 2\
    --gradient_accumulation_steps 8
kill $VLLM_PID
    
```

**Comment by kaiyliu** [comment]
Hi, what version of vllm are you using? I noticed that vllm 0.8.5.post1 provided in the requirements.txt seems not to be correctly invoked. Specifically, there is no corresponding address for vllm, for example, “get_world_size” is an address that is no longer supported.

https://github.com/zxiangx/LC-R1/blob/7a6bd8b8a88d53c04d091ee8100774216f274d27/trl/extras/vllm_client.py#L198

<img width="675" height="258" alt="Image" src="https://github.com/user-attachments/assets/8ac6604c-9723-42dd-a330-c4c35db20f0f" />

**Comment by zxiangx** [comment]
Hello, @LHB-kk! I checked out your issue and tried to reproduce it on my end, but I didn't run into the training getting stuck. On my setup (with 4 A800 GPUs), training Qwen3-4B takes about 7 minutes per step, and it runs normally. To help me dig into why your training might be getting stuck, could you tell me more about the GPU usage on your training cards right after vLLM finishes sampling? That info would be really helpful for troubleshooting.

Also, just a tip if you're training **Qwen3-4B**: you might want to try adjusting the model's tokenizer template. Specifically, try putting the `<think>` token at the end of the prompt. Because the code is based on the `r1-distill-qwen` model, designed to follow that model's template (where the part from the start of the completion to `</think>` is treated as the "thinking" section). Making this adjustment could help your training align better with that framework.

Hope this helps!

**Comment by JinyiHan99** [comment]
> Hello, I am trying to reproduce the experiment, but I found that the training got stuck and there is no progress.
> 
> [2025-07-03 13:32:40,450] [INFO] [config.py:925:print]   zero_enabled ................. False
> [2025-07-03 13:32:40,450] [INFO] [config.py:925:print]   zero_force_ds_cpu_optimizer .. True
> [2025-07-03 13:32:40,450] [INFO] [config.py:925:print]   zero_optimization_stage ...... 0
> [2025-07-03 13:32:40,450] [INFO] [config.py:911:print_user_config]   json = {
>     "train_batch_size": 48, 
>     "train_micro_batch_size_per_gpu": 2, 
>     "gradient_accumulation_steps": 8, 
>     "zero_optimization": {
>         "stage": 0, 
>         "offload_optimizer": {
>             "device": "cpu", 
>             "nvme_path": null
>         }, 
>         "offload_param": {
>             "device": "cpu", 
>             "nvme_path": null
>         }, 
>         "stage3_gather_16bit_weights_on_model_save": false
>     }, 
>     "gradient_clipping": 0.1, 
>     "steps_per_print": inf, 
>     "bf16": {
>         "enabled": true
>     }, 
>     "fp16": {
>         "enabled": false
>     }
> }
> Gradient accumulation steps mismatch: GradientAccumulationPlugin has 1, DeepSpeed config has 8. Using DeepSpeed's value.
> Using /root/.cache/torch_extensions/py311_cu124 as PyTorch extensions root...
> Detected CUDA files, patching ldflags
> Emitting ninja build file /root/.cache/torch_extensions/py311_cu124/cpu_adam/build.ninja...
> /usr/local/lib/python3.11/site-packages/torch/utils/cpp_extension.py:2059: UserWarning: TORCH_CUDA_ARCH_LIST is not set, all archs for visible cards are included for compilation. 
> If this is not desired, please set os.environ['TORCH_CUDA_ARCH_LIST'].
>   warnings.warn(
> Building extension module cpu_adam...
> Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
> ninja: no work to do.
> Loading extension module cpu_adam...
> Time to load cpu_adam op: 2.2426810264587402 seconds
> Using /root/.cache/torch_extensions/py311_cu124 as PyTorch extensions root...
> Detected CUDA files, patching ldflags
> Emitting ninja build file /root/.cache/torch_extensions/py311_cu124/cpu_adam/build.ninja...
> /usr/local/lib/python3.11/site-packages/torch/utils/cpp_extension.py:2059: UserWarning: TORCH_CUDA_ARCH_LIST is not set, all archs for visible cards are included for compilation. 
> If this is not desired, please set os.environ['TORCH_CUDA_ARCH_LIST'].
>   warnings.warn(
> Building extension module cpu_adam...
> Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
> ninja: no work to do.
> Loading extension module cpu_adam...
> Time to load cpu_adam op: 2.2484261989593506 seconds
> Using /root/.cache/torch_extensions/py311_cu124 as PyTorch extensions root...
> Detected CUDA files, patching ldflags
> Emitting ninja build file /root/.cache/torch_extensions/py311_cu124/cpu_adam/build.ninja...
> /usr/local/lib/python3.11/site-packages/torch/utils/cpp_extension.py:2059: UserWarning: TORCH_CUDA_ARCH_LIST is not set, all archs for visible cards are included for compilation. 
> If this is not desired, please set os.environ['TORCH_CUDA_ARCH_LIST'].
>   warnings.warn(
> Building extension module cpu_adam...
> Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
> ninja: no work to do.
> Loading extension module cpu_adam...
> Time to load cpu_adam op: 2.2451608180999756 seconds
>   0%|                                                                                                                                                                | 0/417 [00:00<?, ?it/s
> the vLLM log:
> 
> INFO:     127.0.0.1:59174 - "POST /update_named_param/ HTTP/1.1" 200 OK
> INFO:     127.0.0.1:59174 - "POST /update_named_param/ HTTP/1.1" 200 OK
> INFO:     127.0.0.1:59174 - "POST /update_named_param/ HTTP/1.1" 200 OK
> INFO:     127.0.0.1:59174 - "POST /update_named_param/ HTTP/1.1" 200 OK
> INFO:     127.0.0.1:59174 - "POST /update_named_param/ HTTP/1.1" 200 OK
> INFO 07-03 13:33:05 [block_pool.py:264] Successfully reset prefix cache
> INFO:     127.0.0.1:59174 - "POST /reset_prefix_cache/ HTTP/1.1" 200 OK
> 
> Processed prompts:   0%|          | 0/48 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]
> Processed prompts:  17%|█▋        | 8/48 [00:46<03:51,  5.79s/it, est. speed input: 7.77 toks/s, output: 398.63 toks/s]
> Processed prompts:  33%|███▎      | 16/48 [01:36<03:13,  6.06s/it, est. speed input: 11.05 toks/s, output: 562.18 toks/s]
> Processed prompts:  50%|█████     | 24/48 [01:48<01:35,  3.98s/it, est. speed input: 13.43 toks/s, output: 888.01 toks/s]
> Processed prompts:  67%|██████▋   | 32/48 [01:53<00:42,  2.64s/it, est. speed input: 18.04 toks/s, output: 1193.96 toks/s]
> Processed prompts: 100%|██████████| 48/48 [01:54<00:00,  1.29s/it, est. speed input: 27.75 toks/s, output: 1936.83 toks/s]
> Processed prompts: 100%|██████████| 48/48 [01:54<00:00,  2.38s/it, est. speed input: 27.75 toks/s, output: 1936.83 toks/s]
> INFO:     127.0.0.1:59174 - "POST /generate/ HTTP/1.1" 200 OK
> And my script:
> 
> #!/bin/bash
> MODEL_PATH=Qwen3-4B # replace to your own model path
> SAVE_PATH=qwen3-4B-dpo  # replace to your own save path
> CONFIG_PATH=accelerate_configs/deepspeed_zero2.yaml
> 
> mkdir -p "$SAVE_PATH"
> mkdir -p log
> 
> # run vllm-serve
> export CUDA_VISIBLE_DEVICES=3
> fuser -k 8000/tcp
> trl vllm-serve --model "$MODEL_PATH" --port 8000 > log/vllm.log 2>&1 &
> VLLM_PID=$!
> echo "vLLM server PID: $VLLM_PID"
> 
> accelerate launch --config_file $CONFIG_PATH \
>     train.py \
>     --save_path "$SAVE_PATH" \
>     --model_path "$MODEL_PATH" \
>     --lr 3e-6 \
>     --num_generations 8\
>     --num_train_epochs 1\
>     --run_name lcr1-qwen3-4b-grpo-0702\
>     --per_device_train_batch_size 2\
>     --gradient_accumulation_steps 8
> kill $VLLM_PID

我也遇到了一摸一样的问题，我选择的模型是和论文中描述的模型一样。所以，您最后是怎么解决的呢？
